AX: Remove unnecessary properties AXProperty::{HasHighlighting, SupportsRangeValue} and optimize live region status caching

twilco opened the pull request:
#### b0dbae0788e2980f9cf310a979dfeb3d29ff3d65
<pre>
AX: Remove unnecessary properties AXProperty::{HasHighlighting, SupportsRangeValue} and optimize live region status caching
<a href="https://bugs.webkit.org/show_bug.cgi?id=290505">https://bugs.webkit.org/show_bug.cgi?id=290505</a>
<a href="https://rdar.apple.com/147989745">rdar://147989745</a>

Reviewed by Chris Fleizach.

This commit removes AXProperty::HasHighlighting and AXProperty::SupportsRangeValue. These can be expressed in terms of
already-cached properties. Removing them saves memory and avoids one ancestry walk per object created.

This patch features a couple other optimizations:

- Replace AXProperty::LiveRegionStatus with AXProperty::ExplicitLiveRegionStatus so we don&apos;t cache anything if the live
  regions status is computed by default through role alone, saving memory and making each isolated object cheaper to create.

- Make AccessibilityObject::tagName return a const AtomString&amp; instead of an AtomString to avoid an unnecessary string
  copy at callsites.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasHighlighting const):
(WebCore::AXCoreObject::supportsRangeValue const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::defaultLiveRegionStatusForRole):
(WebCore::AXCoreObject::liveRegionStatus const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::handleRoleChanged):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::headingLevel const):
(WebCore::AccessibilityNodeObject::liveRegionStatus const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::tagName const):
(WebCore::AccessibilityObject::defaultLiveRegionStatusForRole): Deleted.
(WebCore::AccessibilityObject::hasHighlighting const): Deleted.
(WebCore::AccessibilityObject::supportsRangeValue const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::canvasHasFallbackContent const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: <a href="https://commits.webkit.org/292782@main">https://commits.webkit.org/292782@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc6d0e9a257731949e3bc82144a435ce34fcdd9a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16630 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102089 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47536 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25081 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73895 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31103 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100011 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87740 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54231 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12515 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5587 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46863 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104110 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24086 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17565 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82949 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24333 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82343 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20732 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27028 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4571 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17581 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24044 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29195 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23870 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27183 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25443 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth reviewed (commented):
lgtm! Nice savings.

I remember a few months ago I printed all of the attributes for an empty <div> and there were several that were clearly redundant. It looks like you addressed many or most of these here, any idea how many are left?

twilco committed:
AX: Remove unnecessary properties AXProperty::{HasHighlighting, SupportsRangeValue} and optimize live region status caching
https://bugs.webkit.org/show_bug.cgi?id=290505
rdar://147989745

Reviewed by Chris Fleizach.

This commit removes AXProperty::HasHighlighting and AXProperty::SupportsRangeValue. These can be expressed in terms of
already-cached properties. Removing them saves memory and avoids one ancestry walk per object created.

This patch features a couple other optimizations:

- Replace AXProperty::LiveRegionStatus with AXProperty::ExplicitLiveRegionStatus so we don't cache anything if the live
  regions status is computed by default through role alone, saving memory and making each isolated object cheaper to create.

- Make AccessibilityObject::tagName return a const AtomString& instead of an AtomString to avoid an unnecessary string
  copy at callsites.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasHighlighting const):
(WebCore::AXCoreObject::supportsRangeValue const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::defaultLiveRegionStatusForRole):
(WebCore::AXCoreObject::liveRegionStatus const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleLiveRegionCreated):
(WebCore::AXObjectCache::handleRoleChanged):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::headingLevel const):
(WebCore::AccessibilityNodeObject::liveRegionStatus const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::tagName const):
(WebCore::AccessibilityObject::defaultLiveRegionStatusForRole): Deleted.
(WebCore::AccessibilityObject::hasHighlighting const): Deleted.
(WebCore::AccessibilityObject::supportsRangeValue const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::canvasHasFallbackContent const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: https://commits.webkit.org/292782@main
